### PR TITLE
LUCENE-9614: add KnnVectorQuery implementation

### PR DESCRIPTION
**Pull request by msokolov**
This is a first cut at an implementation of a query based on K-nearest neighbors with the vector search being done as part of rewrite(). A couple of quirks:

1. I noticed that scores for the default similarity (Euclidean) had very low precision as they got large. Because of the reverse nature: smaller distances mean higher scores, we need to invert in order to gain compatibility with Lucene search scores. The way we were handling this was to apply an `exp(-distance)` to convert distances to scores. That's theoretically sound, but in practice anything over 100 or so was underflowing to zero and becoming indistinguishable. As a stopgap measure, I changed the behavior so that the scores returned by vector search are allowed to be negative and get set to be `-distance` for the reverse-score (Euclidean distance) case. It's in theory OK for these to be negative as long as they are not directly used as Lucene result scores. I added a further conversion in the Query implementation here that simply adds an offset of the minimum score *for this query*. This is perfectly valid for a single query, but not comparable across queries, and indeed, not even across the same query run on multiple indexes, so it would present problems for distributed implementations. I'm not sure what to do about this yet, and looking for suggestions.
2. There's a clever implementation (hack?!) to deal with trying to minimize over-collection across multiple segments. Basically the idea is to optimistically collect the expected proportion of top K based on the segment size (plus a margin), and then to re-run the query if we can't prove we exhaustively searched the segment. I think it's sound, but welcome comments on that bit since it's a little exotic.

Finally I don't know whether this ought to get pushed without some performance testing; I'll start working on that soon using luceneutil.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        && Arrays.equals(((KnnVectorQuery) obj).target, target);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(field, k, Arrays.hashCode(target));
+  }
+
+  /** Caches the results of a KnnVector search: a list of docs and their scores */
+  class DocAndScoreQuery extends Query {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:173]: nit: make it static so that it doesn't hold a reference to the KnnVectorQuery?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+     */
+    DocAndScoreQuery(int[] docs, float[] scores, int[] segmentStarts) {
+      this.docs = docs;
+      this.scores = scores;
+      this.segmentStarts = segmentStarts;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost)
+        throws IOException {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:197]: Maybe we should do a best-effort check that the reader wrapped by this searcher is effectively the same as the reader that was used for rewriting.

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+            @Override
+            public float getMaxScore(int docid) {
+              docid += context.docBase;
+              float maxScore = 0;
+              for (int idx = Math.max(0, upTo); idx < upper && docs[idx] <= docid; idx++) {
+                maxScore = Math.max(maxScore, scores[idx]);
+              }
+              return maxScore;
+            }
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:254]: let's implement `advanceShallow` as well and make it move `upTo` forward?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+            @Override
+            public float score() {
+              if (upTo >= lower && upTo < upper) {
+                return scores[upTo];
+              }
+              return 0;
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:261]: When can this happen? Since score() should only be called when there is a match, wouldn't it be always legal to return `scores[upTo]`?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+            }
+
+            /**
+             * move the implementation of docIO() into a differently-named method so we can call it
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:265]:
```suggestion
             * move the implementation of docID() into a differently-named method so we can call it
```

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+      if (obj instanceof DocAndScoreQuery == false) {
+        return false;
+      }
+      return Arrays.equals(docs, ((DocAndScoreQuery) obj).docs)
+          && Arrays.equals(scores, ((DocAndScoreQuery) obj).scores);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(Arrays.hashCode(docs), Arrays.hashCode(scores));
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:315]: Can you use the `classHash()` in the hash code too?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+    @Override
+    public String toString(String field) {
+      return KnnVectorQuery.this.toString();
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+      KnnVectorQuery.this.visit(visitor);
+    }
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:302]: Let's give this query its own toString() and visit()?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+            @Override
+            public int docID() {
+              return docIdNoShadow();
+            }
+          };
+        }
+
+        @Override
+        public boolean isCacheable(LeafReaderContext ctx) {
```
Comment (jtibshirani) [inline lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:288]: Is it helpful to cache this, given the expensive part is the rewrite?

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:288]: I think it is. If we returned false here, then any BooleanQuery that has a KnnVectorQuery as a clause would be considered not cacheable, even though it might have other clauses that are actually worth caching.

**Comment by jtibshirani** [comment]
> I noticed that scores for the default similarity (Euclidean) had very low precision as they got large... The way we were handling this was to apply an `exp(-distance)` to convert distances to scores.

I wonder if we could just swap in `f(x) = 1 / (1 + x)`, which decays a lot more slowly than `exp(-x)`. This maintains the nice property producing scores within [0, 1].

> There's a clever implementation (hack?!) to deal with trying to minimize over-collection across multiple segments. Basically the idea is to optimistically collect the expected proportion of top K based on the segment size (plus a margin)...

This is a nice idea! The binomial estimate is based on the idea that nearest vectors are randomly distributed through the index. But since segment membership is related to when a document was indexed, I wonder if it'll be common for most nearest neighbors to be found in one segment. For example, maybe we are indexing (and embedding) news articles as they're written, and our query is a news event. Would it make sense to start with a simple approach where we just collect 'k' from each segment? Then we would explore optimizations in a follow-up with benchmarks?

**Comment by msokolov** [comment]
Thanks for all the comments; I'll follow up with a new commit that addresses them soon. `1 / (1 + x)` makes a lot of sense; I was groping towards it :)

Re: the random-distribution assumption for segments -- I believe this depends very much on the use case. Our experience in e-commerce is it is *usually* true. We've seen occasional outlying cases (more popular media products get re-indexed more often, and there can be correlation if *popularity* is an important query feature, which it is), but this is more the exception than the rule. OTOH a time-series index is likely to be heavily correlated, so a different strategy is appropriate (also, sequential operation can more easily re-use thresholds across segments, and if the segments can be sorted, that will help). Perhaps the vanilla approach (collect K per segment) is best as a safe first step, but I think some optimization here will be heavily impactful since the `K` directly influences the number of nodes explored in the graph, and thence the query cost. Maybe it will deserve some kind of parameterization - so yes, I agree, let's remove this for now, and follow up later.

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java (hunk 4 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:261]: OK; I guess I wasn't totally clear on our ability to make such an assertion. I'll remove the checks and update the test I added for this edge case which I guess doesn't reflect normal usage.

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java (hunk 8 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:288]: I'm not sure; I don't have a lot of experience using the cache, and I guess my idea was that it is *possible* to cache (cacheable)? But ... I guess we only cache the rewritten query? In which case there's really no point since this is already effectively a cache ... yeah I'll change to return false.

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java (hunk 3 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:254]: OK, I am not entirely clear what the contract of this method is, but I'll take a stab. I think it is *not* supposed to update the state of the iterator, but rather return the highest doc after the current one that is <= the argument (like advance)?

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java (hunk 8 above)
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:288]: heh somehow I missed Adrien's last comment ... I retract my previous statement :)

Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:276]: I'm having some trouble following the argument. Seems like a good learning opportunity about caching :)

Even if the BooleanWeight wasn't marked as cacheable, I think its other individual clauses are still eligible for caching. Or is the idea that we'd like to cache the BooleanWeight itself, which may be intersecting the kNN results with an expensive clause? In this case, I guess a takeaway is that a Weight should always be marked cacheable if it's valid, even if it wouldn't actually help performance on its own.

File: lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java [hunk 9]
```diff
@@ -43,9 +43,9 @@ public float compare(float[] v1, float[] v2) {
   };
 
   /**
-   * If true, the scores associated with vector comparisons are in reverse order; that is, lower
-   * scores represent more similar vectors. Otherwise, if false, higher scores represent more
-   * similar vectors.
+   * If true, the scores associated with vector comparisons are nonnegative and in reverse order;
+   * that is, lower scores represent more similar vectors. Otherwise, if false, higher scores
+   * represent more similar vectors, and scores may be negative or positive.
```
[the remark this answers is not in the document]
Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java:48]: So the current patch did not change the `VectorSimilarity.compare` function implementations. It does the `1/1+x` normalization in `VectorsReader.search`, so I think the comment is still valid. I suppose we could impose a requirement on similarity functions that they always return [0,1] in ascending-similar order and eliminate the whole notion of reversed similarities. That would be a nice simplification to the API, and I think we've shown that it is achievable. But if we do decide that's best, I'd like to do it separately since it will touch a bunch of places that aren't changed yet in this PR.

[the remark this answers is not in the document]
Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java:48]: Oh I had misunderstood this docs update. Sounds good to save your idea for a follow-up.

File: lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java (hunk 8 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/KnnVectorQuery.java:276]: This is the mental model I have. {{Weight#isCacheable}} tells whether it's legal to cache the weight, and then it's the job of the cache to tell whether caching is a good idea or not. I believe that one exception to this rule that we make is {{TermInSetQuery}} where we disable caching on clauses that use lots of memory.

**Comment by msokolov** [comment]
I ran a luceneutil test comparing the KnnQuery implementation we have there, which is implemented in createWeight rather than in rewrite and saw no difference. It's a little bit bogus as a comparison perhaps, but it's the best we have right now, and at least it proves we didn't do anything more boneheaded than before.

```
                    TaskQPS baseline      StdDevQPS candidate      StdDev                Pct diff p-value
           LowTermVector      685.43      (6.9%)      664.26      (7.1%)   -3.1% ( -15% -   11%) 0.162
        AndHighMedVector      656.70      (6.1%)      646.58      (2.9%)   -1.5% (  -9% -    7%) 0.308
          HighTermVector      703.33     (11.8%)      704.22      (5.9%)    0.1% ( -15% -   20%) 0.966
        AndHighLowVector      667.55      (7.6%)      669.66      (5.2%)    0.3% ( -11% -   14%) 0.878
                PKLookup      187.46      (1.2%)      188.41      (0.6%)    0.5% (  -1% -    2%) 0.100
           MedTermVector      636.03      (5.9%)      645.88      (5.0%)    1.5% (  -8% -   13%) 0.371
       AndHighHighVector      642.66      (5.8%)      669.80      (6.3%)    4.2% (  -7% -   17%) 0.027
```

By the way I also did try the pro-rating idea I had posted earlier, with mixed results - it consistently made HighTermVector better and MedTermVector worse (quite a bit like 15% less QPS), which really surprised me. But perhaps having a tiny PQ (top K = 1 say) would make the graph exploration quite a bit less efficient? It's also possible this index is skewed and the query is having to re-run a bunch of times ... Needs further investigation.

Finally, I think this is ready to push. I'll push later today if there are no new issues raised.

**Comment by jtibshirani** [comment]
I noticed a couple tiny things after you merged: e48be68, 29ed390.
